File: services/python/shared/middleware/dapr_client.py
"""
Dapr Client for Ag-Fintech Platform
Provides Dapr service mesh operations
"""

import json
import logging
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

import requests

from .idempotency import ProcessedEventsTracker

logger = logging.getLogger(__name__)


@dataclass
class DaprComponents:
    """Available Dapr components"""
    PUBSUB: str = "kafka-pubsub"
    STATE_STORE: str = "redis-state"
    SECRET_STORE: str = "local-secret-store"


@dataclass
class DaprTopics:
    """Available Dapr topics (matching Kafka topics)"""
    FARMER_EVENTS: str = "farmer.events"
    FARM_EVENTS: str = "farm.events"
    CROP_EVENTS: str = "crop.events"
    LIVESTOCK_EVENTS: str = "livestock.events"
    HARVEST_EVENTS: str = "harvest.events"
    EXPENSE_EVENTS: str = "expense.events"
    AUTH_EVENTS: str = "auth.events"
    CACHE_INVALIDATION: str = "cache.invalidation"
    AUDIT_TRAIL: str = "audit.trail"
    NOTIFICATIONS: str = "notifications"
    ANALYTICS: str = "analytics"


class DaprClient:
    """Provides Dapr service mesh operations"""

    def __init__(
        self,
        host: str = "127.0.0.1",
        http_port: str = "3500",
        event_tracker: Optional[ProcessedEventsTracker] = None,
    ):
        self.host = host
        self.http_port = http_port
        self.event_tracker = event_tracker
        self.session = requests.Session()
        self.session.headers.update({"Content-Type": "application/json"})

    @property
    def base_url(self) -> str:
        return f"http://{self.host}:{self.http_port}"

    def publish_event(self, topic: str, data: Any) -> bool:
        """Publish an event via Dapr pub/sub"""
        try:
            url = f"{self.base_url}/v1.0/publish/{DaprComponents.PUBSUB}/{topic}"
            response = self.session.post(url, json=data, timeout=30)

            if response.status_code not in (200, 204):
                logger.error(f"[Dapr] Publish failed with status: {response.status_code}")
                return False

            logger.info(f"[Dapr] Published event to topic: {topic}")
            return True

        except Exception as e:
            logger.error(f"[Dapr] Failed to publish event: {e}")
            # Graceful degradation - don't throw
            return False

    def save_state(
        self,
        key: str,
        value: Any,
        metadata: Optional[Dict[str, str]] = None,
    ) -> bool:
        """Save state via Dapr state management"""
        try:
            url = f"{self.base_url}/v1.0/state/{DaprComponents.STATE_STORE}"
            state_item = [{"key": key, "value": value}]
            if metadata:
                state_item[0]["metadata"] = metadata

            response = self.session.post(url, json=state_item, timeout=30)

            if response.status_code not in (200, 204):
                logger.error(f"[Dapr] Save state failed with status: {response.status_code}")
                return False

            logger.info(f"[Dapr] Saved state: {key}")
            return True

        except Exception as e:
            logger.error(f"[Dapr] Failed to save state: {e}")
            return False

    def save_state_with_etag(
        self,
        key: str,
        value: Any,
        etag: str,
    ) -> bool:
        """Save state with optimistic concurrency (idempotent updates)"""
        try:
            url = f"{self.base_url}/v1.0/state/{DaprComponents.STATE_STORE}"
            state_item = [
                {
                    "key": key,
                    "value": value,
                    "etag": etag,
                    "options": {"concurrency": "first-write"},
                }
            ]

            response = self.session.post(url, json=state_item, timeout=30)

            if response.status_code == 409:
                logger.warning(f"[Dapr] ETag mismatch - state was modified by another process")
                return False

            if response.status_code not in (200, 204):
                logger.error(f"[Dapr] Save state failed with status: {response.status_code}")
                return False

            logger.info(f"[Dapr] Saved state with ETag: {key}")
            return True

        except Exception as e:
            logger.error(f"[Dapr] Failed to save state: {e}")
            return False

    def get_state(self, key: str) -> tuple[Optional[Any], Optional[str]]:
        """Retrieve state via Dapr state management. Returns (value, etag)"""
        try:
            url = f"{self.base_url}/v1.0/state/{DaprComponents.STATE_STORE}/{key}"
            response = self.session.get(url, timeout=30)

            if response.status_code == 404:
                return None, None

            if response.status_code != 200:
                logger.error(f"[Dapr] Get state failed with status: {response.status_code}")
                return None, None

            etag = response.headers.get("ETag")

            try:
                value = response.json()
            except json.JSONDecodeError:
                value = response.text

            logger.info(f"[Dapr] Retrieved state: {key}")
            return value, etag

        except Exception as e:
            logger.error(f"[Dapr] Failed to get state: {e}")
            return None, None

    def delete_state(self, key: str) -> bool:
        """Delete state via Dapr state management"""
        try:
            url = f"{self.base_url}/v1.0/state/{DaprComponents.STATE_STORE}/{key}"
            response = self.session.delete(url, timeout=30)

            if response.status_code not in (200, 204):
                logger.error(f"[Dapr] Delete state failed with status: {response.status_code}")
                return False

            logger.info(f"[Dapr] Deleted state: {key}")
            return True

        except Exception as e:
            logger.error(f"[Dapr] Failed to delete state: {e}")
            return False

    def bulk_get_state(self, keys: List[str]) -> Dict[str, Any]:
        """Retrieve multiple state values"""
        try:
            url = f"{self.base_url}/v1.0/state/{DaprComponents.STATE_STORE}/bulk"
            response = self.session.post(url, json={"keys": keys}, timeout=30)

            if response.status_code != 200:
                logger.error(f"[Dapr] Bulk get state failed with status: {response.status_code}")
                return {}

            results = response.json()
            result_map = {item["key"]: item.get("data") for item in results}

            logger.info(f"[Dapr] Retrieved bulk state: {len(keys)} keys")
            return result_map

        except Exception as e:
            logger.error(f"[Dapr] Failed to bulk get state: {e}")
            return {}

    def invoke_service(
        self,
        service_id: str,
        method_name: str,
        data: Optional[Any] = None,
    ) -> Optional[Any]:
        """Invoke another service via Dapr service invocation"""
        try:
            url = f"{self.base_url}/v1.0/invoke/{service_id}/method/{method_name}"

            if data is not None:
                response = self.session.post(url, json=data, timeout=30)
            else:
                response = self.session.post(url, timeout=30)

            if response.status_code != 200:
                logger.error(f"[Dapr] Service invocation failed with status: {response.status_code}")
                return None

            try:
                result = response.json()
            except json.JSONDecodeError:
                result = response.text

            logger.info(f"[Dapr] Invoked service: {service_id}.{method_name}")
            return result

        except Exception as e:
            logger.error(f"[Dapr] Failed to invoke service: {e}")
            return None

    def get_secret(self, secret_name: str) -> Optional[Dict[str, str]]:
        """Retrieve a secret from Dapr secret store"""
        try:
            url = f"{self.base_url}/v1.0/secrets/{DaprComponents.SECRET_STORE}/{secret_name}"
            response = self.session.get(url, timeout=30)

            if response.status_code != 200:
                logger.error(f"[Dapr] Get secret failed with status: {response.status_code}")
                return None

            logger.info(f"[Dapr] Retrieved secret: {secret_name}")
            return response.json()

        except Exception as e:
            logger.error(f"[Dapr] Failed to get secret: {e}")
            return None

    def check_health(self) -> bool:
        """Check if Dapr sidecar is healthy"""
        try:
            url = f"{self.base_url}/v1.0/healthz"
            response = self.session.get(url, timeout=5)
            return response.status_code in (200, 204)
        except Exception:
            return False

File: server/dapr-client.ts
import { DaprClient, DaprServer, CommunicationProtocolEnum } from '@dapr/dapr';
import { logger } from './logger.js';

const DAPR_HOST = process.env.DAPR_HOST || '127.0.0.1';
const DAPR_HTTP_PORT = process.env.DAPR_HTTP_PORT || '3500';
const DAPR_GRPC_PORT = process.env.DAPR_GRPC_PORT || '50001';

let _daprHealthy = true;
let _lastDaprHealthCheck = 0;
const DAPR_HEALTH_CACHE_MS = 15_000;

logger.info('[Dapr] Initializing', { host: DAPR_HOST, httpPort: DAPR_HTTP_PORT, grpcPort: DAPR_GRPC_PORT });

// Create Dapr client (for outbound calls)
export const daprClient = new DaprClient({
  daprHost: DAPR_HOST,
  daprPort: DAPR_HTTP_PORT,
  communicationProtocol: CommunicationProtocolEnum.HTTP,
});

// Create Dapr server (for inbound calls)
export const daprServer = new DaprServer({
  serverHost: '0.0.0.0',
  serverPort: '3001',
  clientOptions: {
    daprHost: DAPR_HOST,
    daprPort: DAPR_HTTP_PORT,
  },
});

// Dapr component names
export const DAPR_COMPONENTS = {
  PUBSUB: 'kafka-pubsub',
  STATE_STORE: 'redis-state',
  SECRET_STORE: 'local-secret-store',
} as const;

// Dapr topic names (matching Kafka topics)
export const DAPR_TOPICS = {
  FARMER_EVENTS: 'farmer.events',
  FARM_EVENTS: 'farm.events',
  CROP_EVENTS: 'crop.events',
  LIVESTOCK_EVENTS: 'livestock.events',
  HARVEST_EVENTS: 'harvest.events',
  EXPENSE_EVENTS: 'expense.events',
  AUTH_EVENTS: 'auth.events',
  CACHE_INVALIDATION: 'cache.invalidation',
  AUDIT_TRAIL: 'audit.trail',
  NOTIFICATIONS: 'notifications',
  ANALYTICS: 'analytics',
} as const;

/**
 * Publish event via Dapr pub/sub
 */
export async function publishDaprEvent(
  topic: string,
  data: object | string
): Promise<void> {
  try {
    await daprClient.pubsub.publish(DAPR_COMPONENTS.PUBSUB, topic, data);
    _daprHealthy = true;
    logger.debug('[Dapr] Published event', { topic });
  } catch (error) {
    _daprHealthy = false;
    logger.warn('[Dapr] Failed to publish event', { topic, error: (error as Error).message });
  }
}

/**
 * Subscribe to Dapr pub/sub topic
 */
export async function subscribeDaprTopic(
  topic: string,
  handler: (data: unknown) => Promise<void>
): Promise<void> {
  try {
    await daprServer.pubsub.subscribe(
      DAPR_COMPONENTS.PUBSUB,
      topic,
      async (data: unknown) => {
        try {
          await handler(data);
        } catch (error) {
          logger.error(`[Dapr] Error handling event from ${topic}`, { error: (error as Error).message });
        }
      }
    );
    logger.info(`[Dapr] Subscribed to topic: ${topic}`);
  } catch (error) {
    logger.error(`[Dapr] Failed to subscribe to topic ${topic}`, { error: (error as Error).message });
  }
}

/**
 * Save state via Dapr state management
 */
export async function saveState(
  key: string,
  value: unknown,
  metadata?: Record<string, string>
): Promise<void> {
  try {
    await daprClient.state.save(DAPR_COMPONENTS.STATE_STORE, [{ key, value, metadata }]);
    _daprHealthy = true;
  } catch (error) {
    _daprHealthy = false;
    logger.error('[Dapr] Failed to save state', { key, error: (error as Error).message });
    throw error;
  }
}

/**
 * Get state via Dapr state management
 */
export async function getState<T = unknown>(key: string): Promise<T | null> {
  try {
    const response = await daprClient.state.get(DAPR_COMPONENTS.STATE_STORE, key);
    _daprHealthy = true;
    return response as T;
  } catch (error) {
    _daprHealthy = false;
    logger.error('[Dapr] Failed to get state', { key, error: (error as Error).message });
    return null;
  }
}

/**
 * Delete state via Dapr state management
 */
export async function deleteState(key: string): Promise<void> {
  try {
    await daprClient.state.delete(DAPR_COMPONENTS.STATE_STORE, key);
  } catch (error) {
    logger.error('[Dapr] Failed to delete state', { key, error: (error as Error).message });
    throw error;
  }
}

/**
 * Bulk get state via Dapr state management
 */
export async function bulkGetState<T = unknown>(
  keys: string[]
): Promise<Array<{ key: string; value: T | null }>> {
  try {
    const response = await daprClient.state.getBulk(DAPR_COMPONENTS.STATE_STORE, keys);
    _daprHealthy = true;
    return response.map((item) => ({ key: item.key, value: item.data as T }));
  } catch (error) {
    _daprHealthy = false;
    logger.error('[Dapr] Failed to get bulk state', { count: keys.length, error: (error as Error).message });
    return keys.map((key) => ({ key, value: null }));
  }
}

/**
 * Invoke another service via Dapr service invocation
 */
export async function invokeService(
  serviceId: string,
  methodName: string,
  data?: object,
  retries = 2
): Promise<unknown> {
  let lastError: Error | null = null;
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      const response = await daprClient.invoker.invoke(serviceId, methodName, 'post' as any, data);
      _daprHealthy = true;
      return response;
    } catch (error) {
      lastError = error as Error;
      _daprHealthy = false;
      if (attempt < retries) {
        const delay = Math.min(500 * Math.pow(2, attempt), 5000);
        await new Promise((r) => setTimeout(r, delay));
      }
    }
  }
  logger.error(`[Dapr] Failed to invoke service ${serviceId}.${methodName} after ${retries + 1} attempts`, { error: lastError?.message });
  throw lastError;
}

/**
 * Get secret from Dapr secret store
 */
export async function getSecret(
  secretName: string,
  metadata?: Record<string, string>
): Promise<Record<string, string> | null> {
  try {
    const response = await daprClient.secret.get(DAPR_COMPONENTS.SECRET_STORE, secretName, metadata as any);
    return response as Record<string, string>;
  } catch (error) {
    logger.error(`[Dapr] Failed to get secret ${secretName}`, { error: (error as Error).message });
    return null;
  }
}

/**
 * Start Dapr server
 */
export async function startDaprServer(): Promise<void> {
  try {
    await daprServer.start();
    logger.info('[Dapr] Server started');
  } catch (error) {
    logger.error('[Dapr] Failed to start server', { error: (error as Error).message });
    throw error;
  }
}

export async function stopDaprServer(): Promise<void> {
  try {
    await daprServer.stop();
    logger.info('[Dapr] Server stopped');
  } catch (error) {
    logger.error('[Dapr] Failed to stop server', { error: (error as Error).message });
  }
}

export async function checkDaprHealth(): Promise<boolean> {
  if (Date.now() - _lastDaprHealthCheck < DAPR_HEALTH_CACHE_MS) return _daprHealthy;
  try {
    await daprClient.state.get(DAPR_COMPONENTS.STATE_STORE, '_health_check');
    _daprHealthy = true;
    _lastDaprHealthCheck = Date.now();
    return true;
  } catch (err) {
    _daprHealthy = false;
    _lastDaprHealthCheck = Date.now();
    return false;
  }
}

export function isDaprHealthy(): boolean {
  return _daprHealthy;
}
